services/app/src/components/views/quest/QuestEnd.tsx

```diff
@@ -46,48 +216,54 @@ export default class QuestEnd extends React.Component<Props, {}> {
     }
     const tips = [1, 3, 5].map((tip: number) => {
       return (
-        <Button key={tip} onClick={() => this.props.onTip(checkoutError, tip, this.props.quest, this.props.settings, this.state.anonymous, this.state.text, this.state.rating)}>
+        <Button key={tip} onClick={() => this.props.onTip(checkoutError, tip, this.props.quest, this.props.settings, this.state.anonymous, this.formatFeedback(), this.state.rating)}>
           ${tip}
         </Button>
       );
     });
-    // TODO TextField underlineShow={false}
+
+    return (<span>
+      <p>Tip the author (optional):</p>
+      <div className={'tipAmounts ' + (checkoutError === null ? '' : 'checkoutDisabled')}>
+          {tips}
+      </div>
+    </span>);
+  }
+
+  private formatFeedback(): string {
+    if (this.isGoodRating()) {
+      return `Favorite Part: ${this.state.favoritePart || 'not given'}\nDetails: ${this.state.text}`;
+    } else {
+      return `Primary Issue: ${this.state.primaryIssue || 'not given'}\nQualifier: ${this.state.issueQualifier || 'not given'}\nDetails: ${this.state.text}`;
+    }
+  }
+
+  private validateAndSubmit() {
+    if (this.isBadRating() && !this.state.primaryIssue) {
+      this.setState({needsIssue: true});
+      return;
+    } else {
+      this.setState({needsIssue: false});
+    }
+    this.props.onSubmit(this.props.quest, this.props.settings, this.props.user, this.state.anonymous, this.formatFeedback(), this.state.rating);
```

Comment:
Shouldn't this be nested inside the above else, otherwise it'll fire even when invalid?

Reply:
Doesn't strictly matter since we're returning early, but it does make more sense that way. Done.